File: android/core/domain/src/main/kotlin/com/akaar/core/domain/model/Profiles.kt
package com.akaar.core.domain.model

data class UserProfile(
    val id: String,
    val fullName: String?,
    val preferredLang: String,
    val roles: List<AppRole>,
    val activeRole: AppRole?,
) {
    fun has(role: AppRole) = roles.contains(role)
    val needsRoleChoice: Boolean get() = activeRole == null && roles.size > 1
}

/** Buyer-visible. Exact address and legal name are not on this type by design. */
data class ArtisanProfile(
    val userId: String,
    val displayName: String,
    val state: String,
    val district: String,
    val languages: List<String>,
    val experienceYears: Int?,
    val story: Translatable = Translatable(),
    val clusterId: String?,
    val verificationTier: VerificationTier,
) {
    /** District granularity is the finest a buyer needs and the coarsest that stays useful. */
    val regionLabel: String get() = "$district, $state"
}

data class BuyerProfile(
    val userId: String,
    val orgName: String,
    val orgType: BuyerOrgType,
    val city: String?,
    val state: String?,
    val verificationTier: VerificationTier,
)

data class Craft(
    val id: String,
    val slug: String,
    val parentId: String?,
    val nameEn: String,
    val nameHi: String,
    val regions: List<String> = emptyList(),
    val care: Translatable = Translatable(),
) {
    fun name(lang: String) = if (lang == "hi") nameHi else nameEn
}

data class TaxonomyTerm(
    val id: String,
    val slug: String,
    val nameEn: String,
    val nameHi: String,
) {
    fun name(lang: String) = if (lang == "hi") nameHi else nameEn
}

File: mobile/packages/core-domain/src/profiles.ts
import type { AppRole, BuyerOrgType, VerificationTier } from "./enums";
import type { Translatable } from "./translatable";

export interface UserProfile {
  id: string;
  fullName?: string;
  preferredLang: string;
  roles: AppRole[];
  activeRole: AppRole | null;
}

export function has(profile: UserProfile, role: AppRole): boolean {
  return profile.roles.includes(role);
}

export function needsRoleChoice(profile: UserProfile): boolean {
  return profile.activeRole == null && profile.roles.length > 1;
}

/** Buyer-visible. Exact address and legal name are not on this type by design. */
export interface ArtisanProfile {
  userId: string;
  displayName: string;
  state: string;
  district: string;
  languages: string[];
  experienceYears: number | null;
  story?: Translatable;
  clusterId: string | null;
  verificationTier: VerificationTier;
}

/** District granularity is the finest a buyer needs and the coarsest that stays useful. */
export function regionLabel(profile: ArtisanProfile): string {
  return `${profile.district}, ${profile.state}`;
}

export interface BuyerProfile {
  userId: string;
  orgName: string;
  orgType: BuyerOrgType;
  city: string | null;
  state: string | null;
  verificationTier: VerificationTier;
}

export interface Craft {
  id: string;
  slug: string;
  parentId: string | null;
  nameEn: string;
  nameHi: string;
  regions?: string[];
  care?: Translatable;
}

export function craftName(craft: Craft, lang: string): string {
  return lang === "hi" ? craft.nameHi : craft.nameEn;
}

export interface TaxonomyTerm {
  id: string;
  slug: string;
  nameEn: string;
  nameHi: string;
}

export function taxonomyTermName(term: TaxonomyTerm, lang: string): string {
  return lang === "hi" ? term.nameHi : term.nameEn;
}
